File: 04-estructuras-datos/07-tablas-hash/scala/implementaciones/TablaHashEncadenada.scala
import scala.collection.mutable.ArrayBuffer

final class TablaHashEncadenada[K, V](capacidad: Int):
  private val cubetas = Array.fill(capacidad)(ArrayBuffer.empty[(K, V)])

  private def indice(clave: K): Int = Math.floorMod(clave.##, capacidad)

  def put(clave: K, valor: V): Unit =
    val cubeta = cubetas(indice(clave))
    cubeta.indexWhere(_._1 == clave) match
      case -1 => cubeta += ((clave, valor))
      case i  => cubeta(i) = ((clave, valor))

  def get(clave: K): Option[V] =
    // La colisión se resuelve buscando la clave dentro de su cubeta.
    cubetas(indice(clave)).find(_._1 == clave).map(_._2)

File: 04-estructuras-datos/07-tablas-hash/java/implementaciones/TablaHashEncadenada.java
import java.util.LinkedList;
import java.util.List;
import java.util.Objects;

public class TablaHashEncadenada<K, V> {
    private record Entrada<K,V>(K clave, V valor) {}
    private final List<Entrada<K,V>>[] cubetas;

    @SuppressWarnings("unchecked")
    public TablaHashEncadenada(int capacidad) {
        cubetas = (List<Entrada<K,V>>[]) new List<?>[capacidad];
        for (int i = 0; i < capacidad; i++) cubetas[i] = new LinkedList<>();
    }

    private int indice(K clave) {
        // floorMod evita índices negativos cuando el hash de la clave es negativo.
        return Math.floorMod(Objects.hashCode(clave), cubetas.length);
    }

    public void put(K clave, V valor) {
        List<Entrada<K,V>> cubeta = cubetas[indice(clave)];
        cubeta.removeIf(e -> Objects.equals(e.clave(), clave));
        cubeta.add(new Entrada<>(clave, valor));
    }

    public V get(K clave) {
        for (Entrada<K,V> e : cubetas[indice(clave)])
            if (Objects.equals(e.clave(), clave)) return e.valor();
        return null;
    }
}
